### PR TITLE
Extend connection deadline after each multi-get operation

### PR DESCRIPTION
**Pull request by dannykopping**
Fixes https://github.com/grafana/loki/issues/11345

_tl;dr: When a connection is retrieved from the pool, its deadline is extended by the `Timeout` - this works in almost all cases, except `GetMulti` which issues several requests over the connection but the deadline is not extended between each because it reuses the same connection. This leads to requests timing out which should not._

I had to move the connection buffer behind an interface so I could mock out a "slow" network in the `TestGetMultiDeadlineExtension` test.

I don't love the idea of passing `*conn` down, but this was the cleanest approach I could devise that would not necessitate extensive refactoring.

### CODE REVIEW

**Review by 56quarters — COMMENTED** [review]
The core change here, extending the deadline in the parse loop, makes sense to me. 

However, I don't really like:
* Unit tests that rely on timing.
* The changes required to be able to swap out the reader/writer for the unit test.

My preference would be to leave most things as is and just extend the deadline. Since this code as been _mostly_ working for seemingly the entire life of this client and the fix is a single line, I don't think the unit test provides enough value to justify the other changes.

**Comment by dannykopping** [comment]
@56quarters that's fair; at least the test validates that the change works so I'll remove it and the refactoring.
A refactor would be a good idea to make this code more testable, but that can be done separately

**Comment by dannykopping** [comment]
@56quarters done

**Review by 56quarters — APPROVED** [review]
LGTM. Thanks!

File: memcache/memcache.go [hunk 1]
```diff
@@ -653,9 +657,12 @@ func (c *Client) GetMulti(keys []string, opts ...Option) (map[string]*Item, erro
 
 // parseGetResponse reads a GET response from r and calls cb for each
 // read and allocated Item
-func (c *Client) parseGetResponse(r *bufio.Reader, opts *Options, cb func(*Item)) error {
+func (c *Client) parseGetResponse(r *bufio.Reader, conn *conn, opts *Options, cb func(*Item)) error {
 	for {
+		// extend deadline before each additional call, otherwise all cumulative calls use the same overall deadline
+		conn.extendDeadline()
```
Comment (pracucci) [inline memcache/memcache.go:663]: Have you check what's the cost of extending the deadline? May it be impactful?

Comment (dannykopping) [reply memcache/memcache.go:663]: AFAICS it's just modifying a timer [on the connection itself](https://github.com/grafana/gomemcache/blob/6947259a05865b6a7482cb73b556591498f3da0a/memcache/memcache.go#L242).

Under the hood I traced this down to [`setDeadlineImpl`](https://cs.opensource.google/go/go/+/master:src/internal/poll/fd_plan9.go;l=119?q=setdeadline&ss=go%2Fgo), looks fairly inexpensive to me.

Comment (pracucci) [reply memcache/memcache.go:663]: Ok, thanks for checking.
